crates/common/mqtt_channel/src/connection.rs

```diff
@@ -200,8 +204,17 @@ impl Connection {
         mut event_loop: EventLoop,
         mut message_sender: mpsc::UnboundedSender<MqttMessage>,
         mut error_sender: mpsc::UnboundedSender<MqttError>,
+        awaiting_acks: mpsc::UnboundedReceiver<()>,
+        done: oneshot::Sender<()>,
     ) -> Result<(), MqttError> {
+        let awaiting_acks = awaiting_acks.peekable();
+        futures::pin_mut!(awaiting_acks);
         loop {
+            if let Poll::Ready(None) = futures::poll!(awaiting_acks.as_mut().peek()) {
+                // If the channel is dropped, the sender loop has stopped
+                // and we've received an ack for every message published
+                mqtt_client.disconnect().await.unwrap();
```

Comment:
I would rather ignore errors here.

```suggestion
                let _ = mqtt_client.disconnect().await;
```